nexus/db-queries/src/db/datastore/clickhouse_policy.rs

```diff
@@ -0,0 +1,267 @@
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::db::datastore::test_utils::datastore_test;
+    use nexus_inventory::now_db_precision;
+    use nexus_test_utils::db::test_setup_database;
+    use nexus_types::deployment::ClickhouseMode;
+    use omicron_test_utils::dev;
+
+    #[tokio::test]
+    async fn test_clickhouse_policy_basic() {
+        // Setup
+        let logctx = dev::test_setup_log("test_clickhouse_policy_basic");
+        let mut db = test_setup_database(&logctx.log).await;
+        let (opctx, datastore) = datastore_test(&logctx, &db).await;
+
+        // Listing an empty table should return an empty vec
+
+        assert!(datastore
+            .clickhouse_policy_list(&opctx, &DataPageParams::max_page())
+            .await
+            .unwrap()
+            .is_empty());
+
+        // Fail to insert a policy with version 0
+        let mut policy = ClickhousePolicy {
+            version: 0,
+            mode: ClickhouseMode::SingleNodeOnly,
+            time_created: now_db_precision(),
+        };
+
+        assert!(datastore
+            .clickhouse_policy_insert_latest_version(&opctx, &policy)
+            .await
+            .is_err());
+
+        // Inserting version 2 before version 1 should not work
+        policy.version = 2;
+        assert!(datastore
+            .clickhouse_policy_insert_latest_version(&opctx, &policy)
+            .await
+            .is_err());
+
+        // Inserting version 1 should work
+        policy.version = 1;
+        assert!(datastore
+            .clickhouse_policy_insert_latest_version(&opctx, &policy)
+            .await
+            .is_ok());
+
+        // Inserting version 2 should work
+        policy.version = 2;
+        assert!(datastore
+            .clickhouse_policy_insert_latest_version(&opctx, &policy)
+            .await
+            .is_ok());
+
+        // Inserting version 4 should not work, since the prior version is 2
+        policy.version = 4;
+        assert!(datastore
+            .clickhouse_policy_insert_latest_version(&opctx, &policy)
+            .await
+            .is_err());
+
+        // Inserting version 3 should work
+        policy.version = 3;
+        assert!(datastore
+            .clickhouse_policy_insert_latest_version(&opctx, &policy)
+            .await
+            .is_ok());
+
+        // Inserting version 4 should work
+        policy.version = 4;
+        policy.mode =
+            ClickhouseMode::Both { target_servers: 3, target_keepers: 5 };
+        assert!(datastore
+            .clickhouse_policy_insert_latest_version(&opctx, &policy)
+            .await
+            .is_ok());
+
+        let history = datastore
+            .clickhouse_policy_list(&opctx, &DataPageParams::max_page())
+            .await
+            .unwrap();
+
+        for i in 1..=4 {
+            let policy = &history[i - 1];
```

Comment:
Does `paginated` always return ordered rows? I guess it must if you're paging by the thing you expect to be ordered.

Reply:
Yes, and I believe you can change the order of pagination.